src/Datadog.Trace.ClrProfiler.Native/rejit_handler.cpp

```diff
@@ -75,6 +74,79 @@ void RejitHandlerModuleMethod::SetMethodReplacement(const MethodReplacement& met
     m_methodReplacement = std::make_unique<MethodReplacement>(methodReplacement);
 }
 
+void RejitHandlerModuleMethod::RequestRejitForInlinersInModule(ModuleID moduleId)
+{
+    Debug("RejitHandlerModuleMethod::RequestRejitForInlinersInModule: ", moduleId);
+    std::lock_guard<std::mutex> guard(m_ngenModulesLock);
+
+    // We check first if we already processed this module to skip it.
+    auto find_res = m_ngenModules.find(moduleId);
+    if (find_res != m_ngenModules.end())
+    {
+        return;
+    }
+
+    // Enumerate all inliners and request rejit
+    ModuleID currentModuleId = m_module->GetModuleId();
+    mdMethodDef currentMethodDef = m_methodDef;
+    RejitHandler* handler = m_module->GetHandler();
+    ICorProfilerInfo6* pInfo = handler->GetCorProfilerInfo6();
+
+    if (pInfo != nullptr)
+    {
+        // Now we enumerate all methods that inline the current methodDef
+        BOOL incompleteData = false;
+        ICorProfilerMethodEnum* methodEnum;
+
+        HRESULT hr = pInfo->EnumNgenModuleMethodsInliningThisMethod(moduleId, currentModuleId, currentMethodDef,
+                                                                    &incompleteData, &methodEnum);
+        if (SUCCEEDED(hr))
+        {
+            if (!incompleteData)
+            {
+                COR_PRF_METHOD method;
+                unsigned int total = 0;
+                std::vector<ModuleID> modules;
+                std::vector<mdMethodDef> methods;
+                while (methodEnum->Next(1, &method, NULL) == S_OK)
+                {
+                    Debug("NGEN:: Asking rewrite for inliner [ModuleId=", method.moduleId,
+                         ",MethodDef=", method.methodId, "]");
+                    modules.push_back(method.moduleId);
+                    methods.push_back(method.methodId);
+                    total++;
+                }
+                methodEnum->Release();
+                methodEnum = nullptr;
+                if (total > 0)
+                {
+                    handler->EnqueueForRejit(modules, methods);
+                    Info("NGEN:: Processed with ", total, " inliners [ModuleId=", currentModuleId,
+                         ",MethodDef=", currentMethodDef, "]");
+                }
+                m_ngenModules[moduleId] = true;
```

Comment:
On https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo6-enumngenmodulemethodsinliningthismethod-method#remarks, it says that the `incompleteData` result will return true if the enum does not contain all of the caller methods. Should we do the ReJIT regardless of the result and then only save this `m_ngenModules[moduleId] = true` result at the end of this `!incompleteData` block? I'm concerned about a scenario where we always get `incompleteData==true` so we never ReJIT a group of methods that in-line a call we care about

Reply:
Yes, makes sense to me. I wonder if we should track at method level then? Or it doesn’t matter if we call ReJIT multiple times? WDYT?

Reply:
Perhaps we can start by calling ReJIT each time and if we see large overheard we can revisit it?

Reply:
Agreed. Done in ae2b27d